### PR TITLE
Add `TransactionResult` type.

### PR DESCRIPTION
**Pull request by danilvpetrov**
This PR adds `persistence.TransactionResult` type returned by `persistence.Transaction.Commit()` method. This type is the container for all committed event for the committed event and queue items that looks like the following:
```go
// TransactionResult is the result returned by Transaction.Commit() method.
type TransactionResult struct {
	EventItems []*eventstore.Item
	QueueItems []*queuestore.Item
}
```

### CODE REVIEW

File: persistence/transaction_test.go [hunk 1]
```diff
@@ -63,10 +65,61 @@ var _ = Describe("func WithTransaction", func() {
 		Expect(items).NotTo(BeEmpty())
 	})
 
+	It("returns the transaction result", func() {
```
Comment (jmalloc) [inline persistence/transaction_test.go:68]: IMO this test should be moved into the `providertest/eventstore` package.

```suggestion
	It("includes event store items in the transaction result", func() {
```

File: persistence/transaction.go [hunk 2]
```diff
@@ -23,12 +23,19 @@ type Transaction interface {
...
 
 	// Commit applies the changes from the transaction.
-	Commit(ctx context.Context) error
+	Commit(ctx context.Context) (*TransactionResult, error)
 
 	// Rollback aborts the transaction.
 	Rollback() error
 }
 
+// TransactionResult is the result returned by Transaction.Commit() method.
```
Comment (jmalloc) [inline persistence/transaction.go:32]:
```suggestion
// TransactionResult contains information about a successfully committed transaction.
```

File: persistence/transaction.go [hunk 3]
```diff
@@ -23,12 +23,19 @@ type Transaction interface {
...
+type TransactionResult struct {
+	// EventItems is the slice of eventstore.Item persisted within the boundary
+	// of the transaction.
```
Comment (jmalloc) [inline persistence/transaction.go:35]:
```suggestion
	// EventItems contains the events persisted within the transaction.
```

File: persistence/provider/boltdb/transaction.go [hunk 4]
```diff
@@ -11,27 +11,33 @@ import (
 // data stores.
 type transaction struct {
 	ds     *dataStore
+	result *persistence.TransactionResult
```
Comment (jmalloc) [inline persistence/provider/boltdb/transaction.go:14]:
```suggestion
	result persistence.TransactionResult
```

Regardless of whether we choose to retain the pointer return value from `Commit()`, these `result` fields don't need to be a pointer - they're always needed in the happy path.

File: persistence/provider/boltdb/transaction.go [hunk 5]
```diff
@@ -11,27 +11,33 @@ import (
...
-		return err
+		return nil, err
 	}
 
 	if t.actual != nil {
-		return t.actual.Commit()
+		if err := t.actual.Commit(); err != nil {
+			return nil, err
+		}
+
```
Comment (jmalloc) [inline persistence/provider/boltdb/transaction.go:37]: This would get rid of those those uncoverable lines. I wouldn't recommend that if it was something we could reproduce in any sane way or we were responsible for generating the error in our code.

```suggestion
		return &t.result, t.actual.Commit()
```

File: persistence/provider/memory/eventstore.go [hunk 6]
```diff
@@ -19,7 +20,11 @@ func (t *transaction) SaveEvent(
 		return 0, err
 	}
 
-	return t.event.stageSave(&t.ds.db.event, env), nil
+	t.result.EventItems = append(
+		t.result.EventItems,
+	)
+
```
Comment (jmalloc) [inline persistence/provider/memory/eventstore.go:26]: Probably a remnant of an earlier approach.

```suggestion
	t.result.EventItems = append(
		t.result.EventItems,
	)

```

File: persistence/provider/memory/eventstore.go [hunk 7]
```diff
@@ -126,6 +132,8 @@ func (cs *eventStoreChangeSet) stageSave(
 
 	cs.items = append(cs.items, item)
 
+	result.EventItems = append(result.EventItems, item)
```
Comment (jmalloc) [inline persistence/provider/memory/eventstore.go:135]: I think you can handle this entirely in `Commit()`. You have access to the `eventStoreChangeSet` there, so you can just use `cs.items` as `EventItems` in the result. There are a few benefits to doing so:

- this code wont need to change when `SaveEvent()` loses it's offset return value (#168 
- avoid append()'ing
- don't need `transaction.result` field anymore

File: persistence/provider/sql/transaction.go [hunk 8]
```diff
@@ -12,25 +12,30 @@ import (
...
 	if err := t.ds.checkOpen(); err != nil {
-		return err
+		return nil, err
 	}
 
 	if t.actual != nil {
-		return t.actual.Commit()
+		if err := t.actual.Commit(); err != nil {
+			return nil, err
+		}
```
Comment (jmalloc) [inline persistence/provider/sql/transaction.go:35]: As per suggestion for BoltDB:

```suggestion
		return &t.result, t.actual.Commit()
```

File: persistence/provider/sql/transaction.go [hunk 9]
```diff
@@ -12,25 +12,30 @@ import (
 type transaction struct {
 	ds     *dataStore
 	actual *sql.Tx
+	result *persistence.TransactionResult
```
Comment (jmalloc) [inline persistence/provider/sql/transaction.go:15]:
```suggestion
	result persistence.TransactionResult
```

File: persistence/provider/boltdb/transaction.go [hunk 10]
```diff
@@ -68,6 +74,8 @@ func (t *transaction) begin(ctx context.Context) error {
 		t.actual = t.ds.db.Begin(ctx)
 	}
 
+	t.result = &persistence.TransactionResult{}
+
```
Comment (jmalloc) [inline persistence/provider/boltdb/transaction.go:78]:
```suggestion
```

No longer necessary when not a pointer.

File: persistence/provider/boltdb/transaction.go [hunk 11]
```diff
@@ -81,4 +89,5 @@ func (t *transaction) end() {
 	}
 
 	t.ds = nil
+	t.result = nil
```
Comment (jmalloc) [inline persistence/provider/boltdb/transaction.go:92]:
```suggestion
```

No longer necessary when not a pointer.

File: persistence/provider/memory/eventstore.go (hunk 7 above)
Comment (danilvpetrov) [reply persistence/provider/memory/eventstore.go:135]: Cool, i will do this in a follow-up commit.

File: persistence/provider/memory/eventstore.go (hunk 6 above)
Comment (danilvpetrov) [reply persistence/provider/memory/eventstore.go:26]: Hmm, is it one of those suggestions that are the same as the original, GH would not probably like that.

Comment (jmalloc) [reply persistence/provider/memory/eventstore.go:26]: Sorry man! Fixed it. I think there was an extra newline after or something too.

File: persistence/provider/memory/transaction.go [hunk 12]
```diff
@@ -18,27 +18,31 @@ type transaction struct {
...
 	}
 
 	if err := t.ds.checkOpen(); err != nil {
-		return err
+		return nil, err
 	}
 
 	if !t.hasLock {
-		return nil
+		return nil, nil
```
Comment (jmalloc) [inline persistence/provider/memory/transaction.go:35]:
```suggestion
		return &persistence.TransactionResult{}, nil
```

This is a transaction in which nothing happened, so it still needs a valid result, it's just empty. This means we're missing a test-case that an empty `TransactionResult` is returned when no operations are done in the transaction. That should go in `providertest/transaction.go`.

Comment (danilvpetrov) [reply persistence/provider/memory/transaction.go:35]: I would assume that this change will be required not only for in-memory engine, right?

Comment (jmalloc) [reply persistence/provider/memory/transaction.go:35]: I think it's only this one, but just add the test and you'll know :D
